lib/mrtrix3/app.py

```diff
@@ -1101,6 +1101,88 @@ def _is_option_group(self, group):
            not group == self._positionals and \
            group.title not in ( 'options', 'optional arguments' )
 
+  class TypeBoolean:
+    def __call__(self, input_value):
+      true_var = "True"
+      false_var = "False"
+      converted_value = None
+      if input_value.lower() == true_var.lower():
+        converted_value = True
+      elif input_value.lower() == false_var.lower():
+        converted_value = False
+      else:
+        raise argparse.ArgumentTypeError('Entered value is not of type boolean')
+      return converted_value
+
+  class TypeIntegerSequence:
+    def __call__(self, input_value):
+      seq_elements = input_value.split(',')
+      int_list = []
+      for i in seq_elements:
+        try:
+          converted_value = int(i)
+          int_list.append(converted_value)
+        except:
+          raise argparse.ArgumentTypeError('Entered value is not an integer sequence')
+      return int_list
+  
+  class TypeFloatSequence:
+    def __call__(self, input_value):
+      seq_elements = input_value.split(',')
+      float_list = []
+      for i in seq_elements:
+        try:
+          converted_value = float(i)
+          float_list.append(converted_value)
+        except:
+          argparse.ArgumentTypeError('Entered value is not a float sequence')
+      return float_list
+
+  class TypeInputDirectory:
+    def __call__(self, input_value):
+      if not os.path.exists(input_value):
+        raise argparse.ArgumentTypeError(input_value + ' does not exist')
+      elif not os.path.isdir(input_value):
+        raise argparse.ArgumentTypeError(input_value + ' is not a directory')
+      else:
+        return input_value
+
+  class TypeOutputDirectory:
+    def __call__(self, input_value):
+      return input_value
```

Comment:
For a number of output types, check:
https://github.com/MRtrix3/mrtrix3/blob/3.0.4/core/app.cpp#L1181-L1235
From memory, when the `-force` option is used, there's a global "output files will be overwritten", but I think that in some instances there can be individual error messages of "this specific file / directory already exists, and will be overwritten at command completion"

Reply:
You mean a warning message needs to be added. The value of `--force` isn't going to be accessible to the output directory type is it?

Reply:
1.   Sorry yes, individual warning messages for individual items, not errors.
2.   In some cases we do issue an error; specifically attempting to overwrite a pre-existing fixel directory using `-force`. Can lead to unexpected behaviour, so we forbid it.
3.   Regarding knowledge of the force-overwrite status, this I don't know. It depends on the internals of `argparse`. But normally I'd read from `app.FORCE_OVERWRITE` in that case, which only gets set after `argparse.parse_args()` is finished, which means it probably will never be set when those callables are invoked.
     Better, which bears some relation to #1975, would be for the callable to yield some class for which a pre-existing file at that path could be checked at the first instance at which the executing script reads from that path.
4.   Pretty sure that the situation where individual warning messages for overwriting individual files is specific to the Python scripts and doesn't happen in the C++ binaries. So if that no longer happens as a natural consequence of these changes, it's not the end of the world, especially given that up to this point it had to be hard-coded on a per-case basis.

Reply:
Yeah, you won't be able to get it to work from inside the type callable. You would need to add some code to the app I think to loop through all types and find the file/directory types